KAFKA-16691: [connect:transform] Support nested field paths on TimestampConverter

jeqo opened the pull request:
Continue implementation of KIP-821. This time focusing on SMT: TimestampConverter.

Apart from specific SMT changes, it includes additions to SingleFieldPath to support updates to fields and schemas. 

[KAFKA-16691]

---

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

jeqo committed:
KAFKA-16691: Add support to update values and schemas in SingleFieldPath

jeqo committed:
KAFKA-16691: Support for nested structures: TimestampConverter

C0urante commented:
I think this leads to a change in behavior. Right now this test (surprisingly!) passes on trunk:

```java
public class TimestampConverterTest {
    // ...

    @Test
    public void testWithSchemaFieldWithDefaultValue() {
        Map<String, String> config = new HashMap<>();
        config.put(TimestampConverter.TARGET_TYPE_CONFIG, "Timestamp");
        config.put(TimestampConverter.FIELD_CONFIG, "timestamp_field");
        xformValue.configure(config);

        java.util.Date defaultFieldValue = new java.util.Date();
        Schema schema = SchemaBuilder.struct()
                .field(
                        "timestamp_field",
                        Timestamp.builder()
                                .defaultValue(defaultFieldValue)
                                .build()
                );
        Struct value = new Struct(schema)
                .put("timestamp_field", DATE_PLUS_TIME.getTime());

        SourceRecord transformed = xformValue.apply(createRecordWithSchema(schema, value));

        assertEquals(Schema.Type.STRUCT, transformed.valueSchema().type());
        Struct transformedValue = (Struct) transformed.value();
        assertEquals(DATE_PLUS_TIME.getTime(), transformedValue.get("timestamp_field"));
        assertNull(transformedValue.schema().field("timestamp_field").schema().defaultValue());
    }
}
```

We don't propagate default values for the fields we convert. Instead, we automatically substitute in the default value if none is found. This is surprising behavior, and has led to things like [KIP-581](https://cwiki.apache.org/confluence/display/KAFKA/KIP-581%3A+Value+of+optional+null+field+which+has+default+value) and [KIP-1040](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=303794677), but I don't think we should change it in this KIP because it's out of scope and, if necessary, can be touched on in KIP-1040 (which is in discussion at the moment).

jeqo committed:
fixup! KAFKA-16691: Support for nested structures: TimestampConverter

Fallback to not propagating default value in this SMT, based on [1]

[1]: https://github.com/apache/kafka/pull/15893#issuecomment-2102926693

jeqo commented:
@C0urante thanks! Agree, we should leave this change of behavior out of the scope of this PR/KIP. I have returned to the previous behavior, PTAL.

jeqo committed:
fixup! KAFKA-16691: Support for nested structures: TimestampConverter

Fallback to not propagating default value in this SMT, based on [1]

[1]: https://github.com/apache/kafka/pull/15893#issuecomment-2102926693

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -202,6 +204,225 @@ public Object valueFrom(Map<String, Object> map) {
+        }
+
+        return updatedParent;
+    }
```
This seems a bit complex for the `TimestampConverter` use case. It seems like we could get everything we wanted with a simpler, less-flexible approach:

```java
    public Map<String, Object> updateMap(
            Map<String, Object> map,
            Function<Object, Object> update
    ) {
        if (map == null) return null;
        Map<String, Object> result = new HashMap<>(map);

        Map<String, Object> parent = result;
        Map<String, Object> child;
        for (String step : stepsWithoutLast()) {
            child = requireMapOrNull(parent.get(step), "nested field access");
            if (child == null) return null;
            child = new HashMap<>(child);
            parent.put(step, child);
            parent = child;
        }

        Object original = parent.get(lastStep());
        if (original == null) return null;
        Object updated = update.apply(original);
        parent.put(lastStep(), updated);
        return result;
    }
```

which could be used in `TimestampConverter` like this:

```java
    private R applySchemaless(R record) {
        Object rawValue = operatingValue(record);
        if (rawValue == null || config.field.isEmpty()) {
            return newRecord(record, null, convertTimestamp(rawValue));
        } else {
            final Map<String, Object> value = requireMap(rawValue, PURPOSE);
            final Map<String, Object> updatedValue = config.field.updateMap(
                    value,
                    this::convertTimestamp
            );

            if (updatedValue == null) {
                return newRecord(record, null, value);
            } else {
                return newRecord(record, null, updatedValue);
            }
        }
    }
```

jeqo replied:
Thanks! I like the proposed approach. The only change I'm considering to add is to return an intact map if value is not found (the proposal is to return null).
This should make the behavior consistent with update Struct and Schema that do not make this distinction.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -202,6 +204,225 @@ public Object valueFrom(Map<String, Object> map) {
+                        whenFound.apply(
+                            originalValue,
+                            field,
+                            updated,
+                            updateSchema.field(field.name()),
+                            this
+                        );
+                    } else {
+                        if (field.schema().type() == Schema.Type.STRUCT) {
+                            Struct fieldValue = updateValue(
+                                field.schema(),
+                                originalValue.getStruct(field.name()),
+                                updateSchema.field(field.name()).schema(),
+                                step + 1,
+                                whenFound,
+                                whenNotFound,
+                                whenOther
+                            );
+                            updated.put(field.name(), fieldValue);
+                        } else {
+                            // add back to not found and apply others, as only leaf values are updated
+                            found = false;
+                            whenOther.apply(originalValue, field, updated, null, this);
+                        }
+                    }
+                } else {
+                    whenOther.apply(originalValue, field, updated, null, this);
+                }
+            }
+        }
+        if (!found) {
+            whenNotFound.apply(
+                originalValue,
+                null,
+                updated,
+                updateSchema.field(steps.get(step)),
+                this);
+        }
+        return updated;
+    }
```
Just for the sake of the `TimestampConverter` SMT, I don't think we need all this complexity. We could get everything we need without custom interfaces and with fewer arguments by doing something like `public Struct updateStruct(Struct original, Schema updatedSchema, BiFunction<Object, Schema, Object> updateField)` (where the last argument accepts the original field and original schema, and returns the new field).

I've sketched out an implementation but feel free to do your own if you prefer:
```java
public Struct updateStruct(
        Struct struct,
        Schema updatedSchema,
        BiFunction<Object, Schema, Object> update
) {
    return updateStruct(
            struct,
            updatedSchema,
            update,
            steps.get(0),
            steps.subList(1, steps.size())
    );
}

private static Struct updateStruct(
        Struct original,
        Schema updatedSchema,
        BiFunction<Object, Schema, Object> update,
        String currentStep,
        List<String> nextSteps
) {
    if (original == null)
        return null;

    Struct result = new Struct(updatedSchema);
    for (Field field : updatedSchema.fields()) {
        String fieldName = field.name();

        if (fieldName.equals(currentStep)) {
            final Object updatedField;

            // Modify this field
            if (nextSteps.isEmpty()) {
                // This is a leaf node
                Object originalField = original.get(fieldName);
                updatedField = update.apply(
                        originalField,
                        original.schema().field(fieldName).schema()
                );
            } else {
                // We have to go deeper
                Struct originalField = requireStructOrNull(original.get(fieldName), "nested field access");
                updatedField = updateStruct(
                        originalField,
                        field.schema(),
                        update,
                        nextSteps.get(0),
                        nextSteps.subList(1, nextSteps.size())
                );
            }

            result.put(fieldName, updatedField);
        } else {
            // Copy over all other fields from the original to the result
            result.put(fieldName, original.get(fieldName));
        }
    }

    return result;
}
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -202,6 +204,225 @@ public Object valueFrom(Map<String, Object> map) {
+        if (operatingSchema.isOptional()) {
+            builder.optional();
+        }
+        if (operatingSchema.defaultValue() != null) {
+            builder.defaultValue(operatingSchema.defaultValue());
+        }
```
Is this correct? What if the schema we're modifying is for a `Struct`? Wouldn't this cause the new schema to have a default `Struct` value that uses the old schema?

jeqo replied:
Good catch. I'm removing this for the current PR -- we can evaluate if this is needed in following PRs.

C0urante replied:
Hmmm... I think this is definitely required, otherwise we drop default values for nested structs whose children we modify.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -202,6 +204,225 @@ public Object valueFrom(Map<String, Object> map) {
+        StructSchemaUpdater whenOther
+    ) {
+        return updateSchema(originalSchema, baselineSchemaBuilder, 0, whenFound, whenNotFound, whenOther);
+    }
```
Similar thoughts on the method signature here: we don't need custom interfaces (I think something like a `Function<Schema, Schema>` would suffice), and we don't need the `whenNotFound` and `whenOther` arguments (IMO `whenFound` can be named something simpler like `update`).

C0urante reviewed (commented):
Thanks @jeqo. This feels pretty complicated for what we need so far; I don't think any of the custom interfaces are required if we narrow down the scope of the internal API to just what's necessary to implement the intended changes in the `TimestampConverter` SMT.

jeqo committed:
fixup! KAFKA-16691: Support for nested structures: TimestampConverter

Simplify update operations only targeting TimestampConverter

jeqo committed:
chore: snnotate SingleFieldPath as Evolving

 So we may break compatibility in minor releases

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java:
```diff
@@ -383,22 +402,18 @@ private R applyWithSchema(R record) {
-                    builder.optional();
                 if (schema.defaultValue() != null) {
                     Struct updatedDefaultValue = applyValueWithSchema((Struct) schema.defaultValue(), builder);
                     builder.defaultValue(updatedDefaultValue);
```
Won't this line fail now? We haven't defined any fields for the new schema, which should render the updated default value struct invalid for it.

C0urante replied:
Running the entire `TimestampConverterTest` suite with a debugger at this line, it looks like we don't cover it yet. We should probably add that coverage even if the code works as-is.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -39,6 +44,7 @@
  * @see <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-821%3A+Connect+Transforms+support+for+nested+structures">KIP-821</a>
  * @see FieldSyntaxVersion
  */
+@InterfaceStability.Evolving
```
This class isn't part of public API, we shouldn't add this annotation.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -202,6 +208,161 @@ public Object valueFrom(Map<String, Object> map) {
+        }
+
+        Object original = parent.get(lastStep());
+        Object updated = update.apply(original);
+        if (updated != null) {
+            parent.put(lastStep(), updated);
+        }
```
It's a little strange to invoke the update function even if `original` is null. Maybe something like this would be more intuitive?

```suggestion
        if (original == null)
            return map;

        Object updated = update.apply(original);
        if (updated != null) {
            parent.put(lastStep(), updated);
        }
```

Or, if we _really_ want to mirror the exact current behavior of `TimestampConverter` (where `null` is inserted when the field isn't found), we could do something like this:
```suggestion
        Object updated = original != null ? update.apply(original) : null;
        parent.put(lastStep(), updated);
```
